SnapshotLifecycleServiceTests testPolicyCRUD Failure

Oh foo-1 why can't you be foo-2

```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=33, name=Thread-4, state=RUNNABLE, group=TGRP-SnapshotLifecycleServiceTests]
	at __randomizedtesting.SeedInfo.seed([8A2CBE77ADFEBB1F:630A2E905817C3E]:0)
Caused by: java.lang.AssertionError: 
Expected: "foo-2"
     but: was "foo-1"
	at __randomizedtesting.SeedInfo.seed([8A2CBE77ADFEBB1F]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.xpack.slm.SnapshotLifecycleServiceTests.lambda$testPolicyCRUD$10(SnapshotLifecycleServiceTests.java:196)
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/966/console
https://scans.gradle.com/s/wvtal77kwzovq

I had another instance of this in my PR build:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-2/10451/console
https://gradle-enterprise.elastic.co/s/oe6bpog42o5ju/tests/rfsroxnx4sflo-7moiun2tnnjde

It was on our reindex_v2 branch, but I had just merged origin/master some hours before and I doubt the reindex changes could cause this.

Did not reproduce locally using:

```
./gradlew ':x-pack:plugin:ilm:test' --tests "org.elasticsearch.xpack.slm.SnapshotLifecycleServiceTests.testPolicyCRUD" -Dtests.seed=4F8B0454D04C415D -Dtests.security.manager=true -Dtests.locale=sr-Latn-RS -Dtests.timezone=America/Managua -Dcompiler.java=12 -Druntime.java=11
```

Exception:

```
WARNING: Uncaught exception in thread: Thread[Thread-4,5,TGRP-SnapshotLifecycleServiceTests]
java.lang.AssertionError: 
Expected: "foo-2"
     but: was "foo-1"
	at __randomizedtesting.SeedInfo.seed([4F8B0454D04C415D]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.xpack.slm.SnapshotLifecycleServiceTests.lambda$testPolicyCRUD$10(SnapshotLifecycleServiceTests.java:210)
	at org.elasticsearch.xpack.slm.SnapshotLifecycleServiceTests.lambda$testPolicyCRUD$5(SnapshotLifecycleServiceTests.java:160)
	at org.elasticsearch.xpack.slm.SnapshotLifecycleServiceTests$FakeSnapshotTask.triggered(SnapshotLifecycleServiceTests.java:323)
	at org.elasticsearch.xpack.core.scheduler.SchedulerEngine.notifyListeners(SchedulerEngine.java:182)
	at org.elasticsearch.xpack.core.scheduler.SchedulerEngine$ActiveSchedule.run(SchedulerEngine.java:210)
```

--- a/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SnapshotLifecycleServiceTests.java
+++ b/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SnapshotLifecycleServiceTests.java
@@ -31,11 +31,13 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 
+import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
@@ -150,7 +152,6 @@ public class SnapshotLifecycleServiceTests extends ESTestCase {
      * Test new policies getting scheduled correctly, updated policies also being scheduled,
      * and deleted policies having their schedules cancelled.
      */
-    @AwaitsFix( bugUrl = "https://github.com/elastic/elasticsearch/issues/44997")
     public void testPolicyCRUD() throws Exception {
         ClockMock clock = new ClockMock();
         final AtomicInteger triggerCount = new AtomicInteger(0);
@@ -206,13 +207,20 @@ public class SnapshotLifecycleServiceTests extends ESTestCase {
             sls.clusterChanged(event);
             assertThat(sls.getScheduler().scheduledJobIds(), equalTo(Collections.singleton("foo-2")));
 
+            CopyOnWriteArrayList<String> triggeredJobs = new CopyOnWriteArrayList<>();
             trigger.set(e -> {
-                // Make sure the job got updated
-                assertThat(e.getJobName(), equalTo("foo-2"));
+                triggeredJobs.add(e.getJobName());
                 triggerCount.incrementAndGet();
             });
             clock.fastForwardSeconds(1);
 
+            // Let's make sure the job got updated
+            // We don't simply assert that triggeredJobs has one element with value "foo-2" because of a race condition that can see the
+            // list containing <foo-2, foo-1>. This can happen because when we update the policy to version 2 (ie. to foo-2) we will
+            // cancel the existing policy (foo-1) without waiting for the thread executing foo-1 to actually interrupt
+            // (see org.elasticsearch.common.util.concurrent.FutureUtils#cancel) which means that foo-1 could actually get to be
+            // rescheduled and re-run before it is indeed cancelled.
+            assertBusy(() -> assertThat(triggeredJobs, contains("foo-2")));
             assertBusy(() -> assertThat(triggerCount.get(), greaterThan(currentCount)));
 
             final int currentCount2 = triggerCount.get();

--- a/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SnapshotLifecycleServiceTests.java
+++ b/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SnapshotLifecycleServiceTests.java
@@ -37,11 +37,11 @@ import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 
-import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.hasItem;
 
 public class SnapshotLifecycleServiceTests extends ESTestCase {
 
@@ -220,7 +220,7 @@ public class SnapshotLifecycleServiceTests extends ESTestCase {
             // cancel the existing policy (foo-1) without waiting for the thread executing foo-1 to actually interrupt
             // (see org.elasticsearch.common.util.concurrent.FutureUtils#cancel) which means that foo-1 could actually get to be
             // rescheduled and re-run before it is indeed cancelled.
-            assertBusy(() -> assertThat(triggeredJobs, contains("foo-2")));
+            assertBusy(() -> assertThat(triggeredJobs, hasItem("foo-2")));
             assertBusy(() -> assertThat(triggerCount.get(), greaterThan(currentCount)));
 
             final int currentCount2 = triggerCount.get();

